roadmap.md

```diff
@@ -0,0 +1,85 @@
+# Nix is everywhere
+
+Nix is the universal build and [configuration management] tool.
+
+[configuration management]: https://www.sebokwiki.org/wiki/Configuration_Management
+
+Software developers use Nix as a matter of course every day, mostly without even noticing.
+Nix runs trivially, anywhere.
+
+For individuals to large organizations, Nix underpins the entire software supply chain:
+- Developer tooling
+- Build automation
+- Binary distribution
+
+To this end, the Nix team will work towards the following goals.
+
+## Make Nix easy to adopt
+
+  - Well-defined target user base
+      - anyone who wants to manage the complexity of - and build - software
+      - transform bits in a declarative and reproducible way
```

Comment:
I remembered this phrase I coined when trying to describe what got me into Nix:

```suggestion
      - anyone who wants to make computers work reliably in the long run
```

Advantages:
- not limited to software development, but still focused on making computers do something
- captures the reproducibility part without using the technical term

One could play with the "want to" part. To make it softer or broader:

```suggestion
      - anyone who cares for computers to work reliably in the long run
```

```suggestion
      - anyone who has an interest in computers working reliably in the long run
```

To make it stricter and narrower:

```suggestion
      - anyone who needs computers to work reliably in the long run
```

```suggestion
      - anyone who requires computers to work reliably in the long run
```

Reply:
Manage complexity of _software_ is more fitting than making _computers_ work reliably, the latter also needs stuff like easily swappable power supply components etc.

And indeede the value of Nix is in uniform handling of one aspect of complexity of managing software systems, not just reproducibility: I don't care about rebulds from scratch quite as much as about atomic updates and reduction of conflict surface.

(Moreover, for some notions of «run reliably» Nix does nothing with the problem that either you update and get new upstream bugs, or don't and could just as well boot a read-only Ubuntu live image or something)

Reply:
```suggestion
      - anyone who requires software to work reliably both right now *and* in the long run, at scale
```

Reply:
I've thought about the problem some more and it seems that we (or at least me) are not clear on the purpose of defining the Nix audience.
Is it to attract people? That's better handled by the marketing team.
Is it to reject PRs? We should probably judge them on merit instead.
Is it to inform a roadmap? Then define one or two personas to focus on, which aren't the whole audience.
Is it something else? Then please enlighten me.

<!--

I don't think we should exclude small scale or individual users. We don't need to mention scale.

```suggestion
      - anyone who requires software to work reliably both right now *and* in the long run
```

Also, I don't think we restrict the audience to people who care about this single aspect. People can be drawn to Nix for other reasons; for example people who want to simplify their cross compilation toolchain setup, but don't feel like reliability is a solvable problem.

-->

Reply:
>  Is it to reject PRs? We should probably judge them on merit instead.

Ideally, having a vision+roadmap makes it easier to decide which PRs are out of scope, i.e. should be closed regardless of their merit because they would take Nix into a direction we don't want.

So the audience shouldn't be too broad or too vague. I'd rather say "Nix is for software developers / DevOps" than "Nix is for anyone who want reliable software".

Reply:
I doubt that we need to reject PRs based on who a person "is". Making it personal (or about groups) is a recipe for disaster.

> Ideally, having a vision+roadmap makes it easier to decide which PRs are out of scope, i.e. should be closed regardless of their merit because they would take Nix into a direction we don't want.

So then we can define the project's scope and direction instead.

> So the audience shouldn't be too broad or too vague.

This only follows if we assume that we take all of the audience's PRs as orders, which can't be the goal.

Reply:
> Is it to attract people?

@roberth No. It is to inform which of the many possible use cases to elevate into the vision statement.

> Is it to reject PRs? We should probably judge them on merit instead.

Yes, and that merit needs an intended receiver.

> Is it to inform a roadmap? Then define one or two personas to focus on, which aren't the whole audience.

That would indeed be for the roadmap, and a roadmap will always only cover a subset of the vision at a time.

> the audience shouldn't be too broad or too vague.

@edolstra yes, this is why I push for finding a concise and clear-cut phrasing. It must not be too narrow either.

After all, this is simply a decision who we, as maintainers, commit to make Nix useful for, eventually. So maybe we should ask ourselves that question differently.

> I'd rather say "Nix is for software developers / DevOps"

We could say that, and decide if we want to deliberately exclude data scientists for example. But we know that scientists already greatly benefit from using Nix. And while DevOps is something people use Nix for, if Nix had become more popular 15 years prior, that job description may never have emerged in the first place.

That target audience is also to some extent an economic decision: What kind of organisation will support Nix development in the future?

> "Nix is for anyone who want reliable software"

This is explicitly not what I'm proposing, because as you say, that's everyone. Nix already is for people who have some pressing need to rely on software, or work on making it reliable and can make use of Nix as a means to that end.

Okay, here's two proposals for next time:
1. Brainstorm concrete groups which fall into our target audience and those who don't.

   Agree on these two sets, remove everything that is controversial. Maybe an apt description will emerge. Otherwise  leave the two sets as the outcome. We did this partially last time, but did not really deal with the non-audience, and did not do the agreement part. We can still update the vision if someone comes up with an obviously better phrase.

2. Set aside the audience topic, and brainstorm use cases that we clearly agree we want to support.
   T
   As said, this will play into defining user personas, and probably the list we will come up with will inform that.

Reply:
> After all, this is simply a decision who we, as maintainers, commit to make Nix useful for, eventually. So maybe we should ask ourselves that question differently.

This should include our current user base.

> > Is it to attract people?
> 
> No. It is to inform which of the many possible use cases to elevate into the vision statement.

Good. Publishing a decision about audience is going to be broad or divisive. Let's not publish one.

> > Is it to reject PRs? We should probably judge them on merit instead.
> 
> Yes, and that merit needs an intended receiver.

Our current user base would be a good starting point. We know quite well who they are.

> > Is it to inform a roadmap? Then define one or two personas to focus on, which aren't the whole audience.
> 
> That would indeed be for the roadmap, and a roadmap will always only cover a subset of the vision at a time.

So, kind of, but not really?

> After all, this is simply a decision who we, as maintainers, commit to make Nix useful for, eventually. So maybe we should ask ourselves that question differently.

I don't think picking favorites in public is something you do "simply".

> > I'd rather say "Nix is for software developers / DevOps"
> 
> We could say that, and decide if we want to deliberately exclude data scientists for example. But we know that scientists already greatly benefit from using Nix. And while DevOps is something people use Nix for, if Nix had become more popular 15 years prior, that job description may never have emerged in the first place.
> 
> That target audience is also to some extent an economic decision: What kind of organisation will support Nix development in the future?

Let's not make badly informed economic decisions.

> I'm fine with just saying "software developers". Anyone else may not have or want to use Nix directly anyway.

We do support indirect use of Nix though, via `--json` for example. Should we exclude them? I think not. Projects that facilitate such use are very committed and if successful they'll want to support the project because it's an essential part of theirs.

> > "Nix is for anyone who want reliable software"
> 
> This is explicitly not what I'm proposing, because as you say, that's everyone. Nix already is for people who have some pressing need to rely on software, or work on making it reliable and can make use of Nix as a means to that end.
> 
> Okay, here's two proposals for next time:
> 
> 1. Brainstorm concrete groups which fall into our target audience and those who don't.
>    Agree on these two sets, remove everything that is controversial. Maybe an apt description will emerge. Otherwise  leave the two sets as the outcome. We did this partially last time, but did not really deal with the non-audience, and did not do the agreement part. We can still update the vision if someone comes up with an obviously better phrase.

I'm not convinced that there's a real need for this.

> 2. Set aside the audience topic, and brainstorm major use cases.
>    Leave all that we clearly agree on, and prioritise roughly.
>    As said, this will play into defining user personas, and probably the list we will come up with will inform that.

Sounds better.

I'd like to see us work toward practically achievable goals and show the community that we're making real progress.

Reply:
I think it's fairly hard to write a vision statement for Nix as a whole, but much easier to rewrite ones for specific layers like libnixstore and libnixexpr. This should be a sign that divvying up responsibilities should make Nix project management easier and avoid clashing opinions or uncertainty.

Reply:
I agree with everything you wrote. Just one note, so we don't misunderstand each other.

> I don't think picking favorites in public is something you do "simply".

I don't see how any of this is about picking favorites. The "simple" here was supposed to refer to what we talked about last week: It should be simple to answer for users or contributors if their intent to use or modify Nix is covered by the vision statement. That is primarily to communicate what kind of support to expect from maintainers.

We should by all means lead the discussion in public, to make sure we're on the right track. That is naturally not simple or easy, because it means saying "no" to some people in order to be able to clearly say "yes" to other people. It always hurts.

At the very least, the rationale behind that decision must be absolutely transparent, in my opinion.

I'm also fine with not determining this and instead skipping to user stories, but I suspect that will amount roughly to the same thing. Nix will not be able to solve everyone's problems, and we have to focus on something in the long term to be able to pick something to focus on in the short term.

Reply:
> It should be simple to answer for users or contributors if their intent to use or modify Nix is covered by the vision statement.

Stated like this, it is «what Nix does» not «what use cases Nix serves» — which is of course better.

(And of course Nix does not help with reliability of software itself, but it is a good tool to reduce cross-contamination of different parts when such interactions are undesirable)

Reply:
> Stated like this, it is «what Nix does» not «what use cases Nix serves»

@7c6f434c It sounds like we are thinking alike :) I started writing a bunch on what I am calling "Supply-" versus "demand-driven" project management.

Reply:
I mean, I myself am more or less Unix-maximalist (composable scope-able predictable tools, all that). Also I sincerely believe that there are many more tools useful _despite_ their intended use cases but thanks to their predictability and friendliness to integrating into larger toolchains — than tools that are actually mainly useful as intended.